buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java

```diff
@@ -0,0 +1,141 @@
+ */
+public class ReaperPlugin implements Plugin<Project> {
+
+    private Project project;
+    private Path baseDirectory;
+    private volatile Process reaperProcess;
+
+    @Override
+    public void apply(Project project) {
+        this.project = project;
+        this.baseDirectory = project.getRootDir().toPath().resolve(".gradle")
+            .resolve("reaper").resolve("build-" + ProcessHandle.current().pid());
+        project.getGradle().addListener(new BuildListener() {
+            @Override
+            public void buildStarted(Gradle gradle) {}
+
+            @Override
+            public void settingsEvaluated(Settings settings) {}
+
+            @Override
+            public void projectsLoaded(Gradle gradle) {}
+
+            @Override
+            public void projectsEvaluated(Gradle gradle) {}
+
+            @Override
+            public void buildFinished(BuildResult result) {
+                if (reaperProcess != null) {
+                    try {
+                        reaperProcess.getOutputStream().close();
+                        project.getLogger().info("Waiting for reaper to exit normally");
+                        if (reaperProcess.waitFor() != 0) {
+                            throw new GradleException("Reaper process failed. Check log at "
+                                + baseDirectory.resolve("error.log") + " for details");
+                        }
+                    } catch (Exception e) {
+                        throw new RuntimeException(e);
+                    }
+
+                }
+            }
+        });
+    }
+
+    /**
+     * Register a pid that will be killed by the reaper.
+     */
+    public void registerPid(String serviceId, long pid) {
+        String[] killPidCommand = OS.<String[]>conditional()
+            .onWindows(() -> new String[]{"Taskill", "/F", "/PID", String.valueOf(pid)})
+            .onUnix(() -> new String[]{"kill", "-9", String.valueOf(pid)}).supply();
+        registerCommand(serviceId, killPidCommand);
+    }
+
+    /**
+     * Register a system command that will be run by the reaper.
+     */
+    public void registerCommand(String serviceId, String... command) {
+        ensureReaperStarted();
+
+        try {
+            Files.writeString(baseDirectory.resolve(serviceId), String.join(" ", command));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public void unregister(String serviceId) {
+        try {
+            Files.deleteIfExists(baseDirectory.resolve(serviceId));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private synchronized void ensureReaperStarted() {
+        if (reaperProcess == null) {
+            try {
+                // copy the reaper jar
+                Path jarPath = project.getBuildDir().toPath().resolve("reaper").resolve("reaper.jar");
```

Comment:
We should look at ways of caching this. As far as I can tell we copy this jar on every build regardless of if the file exists. That said, we can't just check if the file exists as that doesn't mean it's up to date so we'll have to implement our own caching/up-to-date mechanism here.

Reply:
Is it that big a deal? If we are starting the reaper, we are doing much more heavy io from eg starting an elasticsearch cluster, which involves tons of copying. The simplicity of always copying seems nice here, at least for now until it is a bottleneck? Remember we only do this in the case we are starting the reaper, which happens when we are actually launching our first elasticsearch node in testclusters.

Reply:
_Probably_ not, but this is blocking I/O in a `synchronized` method which means we can't proceed to start anymore nodes until this bit is done. We are probably talking milliseconds here but it's something to think about.

Reply:
Milliseconds for the very first cluster started. Given that the vast majority of our test clusters are single node, I don't think parallelism is hurt, and these clusters are very heavyweight. Can we defer this until it is a problem?

Reply:
Agreed. This isn't a hotspot in our build.